### ISSUE TITLE
error fetching info from OAuth2 provider

### ISSUE THREAD

**Issue by nuratun** [realm/realm-js#4995]
### How frequently does the bug occur?

All the time

### Description

We've recently upgraded from Stitch to Realm and authenticating to Google using the official docs. 

Here's the relevant code:

```
useGoogleLogin({
    onSuccess: credentialResponse => {
      const token = credentialResponse.access_token;
      const credentials = Realm.Credentials.google({ token });
      app.logIn(credentials).then((user) => {
        console.log(`Logged in with id: ${user.id}`);
      });
    }, 
onFailure: () => {
      // removed, as irrelevant
    },
});
```

Everything form Google is working fine; I receive the id_token, then I pass it along to Realm. That's when I get the error message.

I've also tried passing along the auth code, but Realm just stalls with nothing returned. If I pass along the access_token, I get the same error as before.

### Stacktrace & log output

```shell
Our app responds with the following after successfully retrieving the id_token and access_code from Google:

Failed to load resource: the server responded with a status of 401 ()

Our logs in MongoDB shows the following:
Error:

error fetching info from OAuth2 provider
Remote IP Address:
[IP address removed for security reasons]
```

### Can you reproduce the bug?

Yes, always

### Reproduction Steps

Install Realm-web:
`npm i --save realm-web` (version currently installed: 1.7.0)

Import it:
`import * as Realm from "realm-web";`

Implement the code found [here](https://www.mongodb.com/docs/realm/sdk/react-native/examples/authenticate-users/#google-user)

Retrieve the Google id_token (or auth code or access_token... doesn't seem to matter). Pass it along:

```
const credentials = Realm.Credentials.google({ idToken });
--
app.logIn(credentials).then((user) => {
console.log(`Logged in with id: ${user.id}`);
});
```

Results in the the same error response every single time: `error fetching info from OAuth2 provider`

If I switch on OpenID, it suddenly works, however, OpenID doesn't return the email, which is necessary for my use case, so that's not useful. 

### Version

1.7.0

### What SDK flavour are you using?

Atlas App Services (auth, functions, etc.)

### Are you using encryption?

Yes, using encryption

### Platform OS and version(s)

Windows, Mac, Linux

### Build environment

Which debugger for React Native: ..

**Comment by nuratun** [comment]
> @nuratun
> 
> When calling `GoogleAuth.initialize()`, did you set `grantOfflineAccess` to false or true?

False, but I can switch it to true when I get back to my computer. I'll let you know the results.

**Comment by nuratun** [comment]
@kneth I finally got back to my computer and tried settings the grantOfflineAccess to true. It returned a code, instead of a tokenId, so I sent that to `await app.logIn( Realm.Credentials.google({ code }) );`

But I still get a 401 error and the error logs on MongoDB reports the same error message: `error fetching info from OAuth2 provider`

Is there anything else I should be doing?

**Comment by nuratun** [comment]
Thanks @kneth. I have seen that thread previously. As I said, the error goes away when I switch to OpenId, but that doesn't help me, as I need the email address returned to identify our users, and OpenId just returns the name of the user logging in. I suppose there's no easy solution for this.

**Comment by desistefanova** [comment]
Hi @nuratun,
It seems that we will need time to investigate the issue with google login.
Until then, you can switch your Atlas app "Authentication Providers" to use "Custom JWT Authentication". You can login using the same google `id_token` if you set the following configurations:

- Verification Method: "Use a JWK URI"
- JWK URI: "https://www.googleapis.com/oauth2/v3/certs"
- Metadata Fields: email:name; email:email
- Audience: "XXXXXXXXXX.apps.googleusercontent.com" (if you don't know the audience for your google app you can get it from "aud" field inside the id_token payload. You can use https://jwt.io to see the payload)

Here is how the configuration in the Atlas app should looks like:
![image](https://user-images.githubusercontent.com/95419820/197272591-fb0c5388-a357-4a13-b1ab-b6669b2eb472.png)

From your code you have to create the credentials using:
 `Realm.Credentials.jwt({ token });`

I hope you can use this workaround until we find a solution for Atlas Google authentication provider.

**Comment by nuratun** [comment]
@desistefanova Thank you so much! Yes, that finally worked. It'll be great if you guys can look into this bug, but since this is a workaround, there's no rush. 

I'm not sure if this bug report should be kept open or closed. I'll leave the decision to you.

**Comment by kneth** [comment]
@nuratun I'll keep the issue open until we have a chance to revisit auth in the near future.

**Comment by ahmed-shehata** [comment]
As a workaround for anyone reading this, you can decode the JWT token using [jwt-decode](https://www.npmjs.com/package/jwt-decode) from Google OAuth login to extract the required fields such as email and name:

```js
import jwtDecode from 'jwt-decode';
//...
const googleData = jwtDecode(idToken);
//... 
const credentials = Realm.Credentials.google({ idToken });
app.logIn(credentials).then((user) => {
console.log(`Logged in with id: ${user.id}`);
console.log("User email: ", googleData.email);
});

```

Where `googleData` is

```js
{
  "at_hash": "<string>",
  "aud": "<string>",
  "azp": "<string>",
  "email": "<string>",
  "email_verified": <boolean>,
  "exp": <timestamp>,
  "family_name": "<string>",
  "given_name": "<string>",
  "hd": "<string>",
  "iat": <timestamp>,
  "iss": "<string>",
  "locale": "<string>",
  "name": "<string>",
  "picture": "<string>",
  "sub": "<string>"
}

```

**Comment by jasontlouro** [comment]
> I hope you can use this workaround until we find a solution
@desistefanova @kneth how is this a workaround? Isn't this a one-way street, since there's no way to migrate user accounts from the Custom JWT method back to the Google method after they're created? 

Searching for a way to unify various providers since custom JWT only lets us handle one (in our case Google) and not many (Google on two apps, Apple on two apps, etc.). Cross-platform auth doesn't seem to be handled at all by Realm/App Services.

**Comment by jamalsoueidan** [comment]
Same issue with Facebook, 2 years and no fix?
